lib/calendar.ts

```diff
@@ -1509,21 +1424,28 @@ const helperHebrew: HelperSharedImpl = ObjectAssign({}, helperSharedImpl, {
     Av: { leap: 12, regular: 11, monthCode: 'M11', days: 30 },
     Elul: { leap: 13, regular: 12, monthCode: 'M12', days: 29 }
   },
-  getMonthCode(year, month) {
+  getMonthCode(year: number, month: number) {
     if (this.inLeapYear({ year })) {
       return month === 6 ? buildMonthCode(5, true) : buildMonthCode(month < 6 ? month : month - 1);
     } else {
       return buildMonthCode(month);
     }
   },
   adjustCalendarDate(
-    this: typeof helperHebrew & HelperSharedImpl,
-    calendarDate,
-    cache,
-    overflow = 'constrain',
+    this: HelperPerCalendarImpl & { months: HebrewMonthInfo; getMonthCode(year: number, month: number): string },
+    calendarDate: Partial<FullCalendarDate>,
+    cache?: OneObjectCache,
+    overflow: Overflow = 'constrain',
     fromLegacyDate = false
   ) {
-    let { year, eraYear, month, monthCode, day, monthExtra } = calendarDate;
+    // The incoming type is actually CalendarDate (same as args to
+    // Calendar.dateFromParams) but TS isn't smart enough to follow all the
+    // reassignments below, so as an alternative to 10+ type casts, we'll lie
+    // here and claim that the type has `day` and `year` filled in already.
+    let { year, eraYear, month, monthCode, day, monthExtra } = calendarDate as Omit<
```

Comment:
Wouldn't this make TypeScript think that the check below (`if (year === undefined ...)`) was redundant? Should the type here be unioned with `{ year?: number, day?: number }` instead?